README.md
# pam-sys - Rust FFI bindings to the Linux Pluggable Authentication Modules (PAM)

[![Crates.io](https://img.shields.io/crates/v/pam-sys.svg)](https://crates.io/crates/pam-sys)
[![Documentation](https://docs.rs/pam-sys/badge.svg)](https://docs.rs/pam-sys/)
[![Build Status Azure](https://dev.azure.com/1wilkens/ci/_apis/build/status/pam-sys?branchName=master)](https://dev.azure.com/1wilkens/ci/_build/latest?definitionId=1&branchName=master)
[![Build Status Cirrus](https://img.shields.io/cirrus/github/1wilkens/pam-sys/master?style=flat-square)](https://cirrus-ci.com/github/1wilkens/pam-sys)
[![License](https://img.shields.io/crates/l/pam-sys.svg?branch=master)](https://travis-ci.org/1wilkens/pam-sys)

This crate uses [`bindgen`](https://github.com/rust-lang/rust-bindgen) to generate the raw FFI
definitions for PAM. For a rustified API consider using [`pam`](https://github.com/1wilkens/pam).

## Supported Rust versions (MSRV)
The library is only continuously built against Rust stable, beta and nightly but as it does not use
a lot of new language features it should probably compile on older versions as well. The MSRV is
mostly determined by the version of `bindgen` used. If you encounter problems building on older
versions and a small fix can be applied to make the build succeed, consider opening a pull request.

## Supported operating systems
Currently this library is only thoroughly tested on Linux (due to the lack of reliable access to
other platforms). However, support of all operating systems supported by PAM is an explicit goal.
If you spot an error, please open an issue and I'll try to resolve it.

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in
the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

<!--
### Added - for new features.
### Changed - for changes in existing functionality.
### Deprecated - for once-stable features removed in upcoming releases.
### Removed - for deprecated features removed in this release.
### Fixed - for any bug fixes.
### Security - to invite users to upgrade in case of vulnerabilities.
-->

## [Unreleased]

## [1.0.0-alpha5] - 2023-11-01
### Changed
- Blocklist `__gnuc_va_list` to fix build on some Linux systems
- Bump `bindgen` to `0.68` and disable default features to reduce dependency footprint (credits to @zombiepigdragon)

## [1.0.0-alpha4] - 2022-01-19
### Added
- Explicitly include `security/openpam.h` on FreeBSD
- Add CI for FreeBSD via cirrus-ci

### Changed
- Slightly optimize `include` setting in `Cargo.toml`
- Split up platform-specific adaptions in `build.rs` to handle FreeBSD/NetBSD with OpenPAM
- Bump bindgen to 0.59 (and switch from deprecated functions)

## [1.0.0-alpha3] - 2021-03-28
### Changed
- Depend on `bindgen` ^0.55 to force macro constants to be generated as signed integers
- Don't derive `Copy` for `PamHandle`

## [1.0.0-alpha2] - 2020-05-03
### Changed
- Pin `bindgen` version for now to avoid accidental changes in the bindings
- Move CI to azure pipelines (and remove `.travis.yml`)

### Removed
- Remove module functions `pam_sm_*` as these are not supposed to be called but rather implemented by modules

## [1.0.0-alpha1] - 2019-11-12
### Changed
- Complete rewrite based on `bindgen`
    - Only expose raw APIs as is convention
    - Easier multi-platform handling
- Update to rust edition 2018
- Update `libc` dependency (0.2.39 -> ^0.2)
- Fix `clippy` lints

## [0.5.6] - 2018-07-04
### Fixed
- Fixed NULL pointer usage in wrapped::get_env function when pam_getenv fails

## [0.5.5] - 2018-03-21
### Added
- Implement `pam_get_user`

### Changed
- Added `cache: cargo` directive to speedup CI
- Updated `libc` dependency (0.2.33 -> 0.2.39)

## [0.5.4] - 2017-11-30
### Changed
- Only provide official support for Rust stable, beta and nightly (mainly through travis)
- Updated `libc` dependency (0.2.20 -> 0.2.33)

### Fixed
- Only link `pam_misc` on Linux

## [0.5.3] - 2017-02-18
### Fixed
- Fixed `readonly` flag in `pam_misc_setenv`
- Added some `null` checks for some `wrapped` functions

## [0.5.2] - 2017-02-18
### Fixed
- Fixed missing link argument to `pam_misc`

## [0.5.1] - 2017-02-17
### Fixed
- Fixed compilation failure on Rust 1.4.0
- Fixed wrong category

## [0.5.0] - 2017-02-17
### Added
- Added travis-ci badge to `Cargo.toml`
- Added categories to `Cargo.toml`

### Changed
- Changed `raw::pam_getenvlist`'s return type back to `*const *const c_char`
- Made `wrapped` functions safe and changed arguments to use better types

### Fixed
- Fixed `raw::pam_misc_dropenv` and `wrapped::misc_dropenv`'s return type

## [0.4.3] - 2017-01-20
### Added
- Added license badge to `README.md`

### Changed
- Updated `libc` dependency (0.2.9 -> 0.2.20)
- Moved documentation to [docs.rs](https://docs.rs/pam-sys/)

### Removed
- Removed obsolete dependencies (`gcc` & `pkg-config`)
- Removed `.travis-update-gh-pages.sh` and obsolete rust versions from `.travis.yml`

## [0.4.1] - 2016-04-11
### Changed
- Relicense under MIT/Apache-2.0

## [0.4.0] - 2016-04-11
### Added
- Better travis-ci integration
    - Test on 1.4.0-1.6.0, stable, beta and nightly
    - Use containers for faster builds

### Changed
- Updated `libc` dependency (0.2.2 -> 0.2.9)
- Changed wrapped/{strerror, getenv} to return `Option<&'static str>` instead of `*const c_char`

## [0.3.0] - 2015-12-08
### Added
- CHANGELOG.md

### Changed
- Updated `libc` dependency (0.1.8 -> 0.2.2)
- PamHandle from empty struct to zero variant enum (as recommended in [the Rust Book](https://doc.rust-lang.org/nightly/book/ffi.html#representing-opaque-structs))


[Unreleased]: https://github.com/1wilkens/pam-sys/compare/v1.0.0-alpha5...HEAD
[1.0.0-alpha5]: https://github.com/1wilkens/pam-sys/compare/v1.0.0-alpha4...v1.0.0-alpha5
[1.0.0-alpha4]: https://github.com/1wilkens/pam-sys/compare/v1.0.0-alpha3...v1.0.0-alpha4
[1.0.0-alpha3]: https://github.com/1wilkens/pam-sys/compare/v1.0.0-alpha2...v1.0.0-alpha3
[1.0.0-alpha2]: https://github.com/1wilkens/pam-sys/compare/v1.0.0-alpha1...v1.0.0-alpha2
[1.0.0-alpha1]: https://github.com/1wilkens/pam-sys/compare/v0.5.6...v1.0.0-alpha1
[0.5.6]: https://github.com/1wilkens/pam-sys/compare/v0.5.5...v0.5.6
[0.5.5]: https://github.com/1wilkens/pam-sys/compare/v0.5.4...v0.5.5
[0.5.4]: https://github.com/1wilkens/pam-sys/compare/v0.5.3...v0.5.4
[0.5.3]: https://github.com/1wilkens/pam-sys/compare/v0.5.2...v0.5.3
[0.5.2]: https://github.com/1wilkens/pam-sys/compare/v0.5.1...v0.5.2
[0.5.1]: https://github.com/1wilkens/pam-sys/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/1wilkens/pam-sys/compare/v0.4.3...v0.5.0
[0.4.3]: https://github.com/1wilkens/pam-sys/compare/v0.4.1...v0.4.3
[0.4.1]: https://github.com/1wilkens/pam-sys/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/1wilkens/pam-sys/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/1wilkens/pam-sys/compare/f051f14b76ad1e06be1832604e0ca570743460ac...v0.3.0


src/types.rs
//! Types defined by Linux-PAM
//!
//! This modules contains struct and enum definitions used by `pam-sys`.

use libc::{c_char, c_int, c_void};

use std::fmt::{Display, Error, Formatter};

/// Type alias for the PAM "conversation function" used as part of the `PamConversation` struct
pub type ConvClosure = (extern "C" fn(c_int,
                                  *mut *mut PamMessage,
                                  *mut *mut PamResponse,
                                  *mut c_void)
                                  -> c_int);

/// Opaque struct internal to Linux-PAM
///
/// From `_pam_types.h`:
///
/// "This is a blind structure. Users aren't allowed to see
/// inside a `pam_handle_t`, so we don't define struct `pam_handle` here.
/// This is defined in a file private to the PAM library.
/// (i.e., it's private to PAM service modules, too!)"
pub enum PamHandle {}

/// Message struct to transfer authentication data to the user
///
/// From `_pam_types.h`:
///
/// "Used to pass prompting text, error messages, or other informatory text to the user.
/// This structure is allocated and freed by the PAM library (or loaded module)."
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct PamMessage {
    pub msg_style: c_int,
    pub msg: *const c_char,
}


/// Response struct to transfer the user's response back to Linux-PAM
///
/// From `_pam_types.h`:
///
/// "Used to return the user's response to the PAM library.
/// This structure is allocated by the application program,
/// and free()'d by the Linux-PAM library (or calling module)."
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct PamResponse {
    pub resp: *mut c_char,
    /// currently un-used, zero expected
    pub resp_retcode: c_int,
}

/// Conversation structure containing the `converse` function and authentication data
///
/// From `_pam_types.h`:
///
/// "The actual conversation structure itself"
#[repr(C)]
pub struct PamConversation {
    /* int (*conv)(int num_msg, const struct pam_message **msg, struct pam_response **resp, void *appdata_ptr); */
    pub conv: Option<ConvClosure>,
    pub data_ptr: *mut c_void,
}

/// Special struct for the `PAM_XAUTHDATA` pam item
///
/// From `_pam_types.h`:
///
/// "Used by the `PAM_XAUTHDATA` pam item. Contains X authentication
/// data used by modules to connect to the user's X display.
/// Note: this structure is intentionally compatible with `xcb_auth_info_t`."
#[repr(C)]
#[derive(Clone, Copy, Debug)]
pub struct PamXAuthData {
    pub namelen: c_int,
    pub name: *mut c_char,
    pub datalen: c_int,
    pub data: *mut c_char,
}

/// The Linux-PAM return values
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum PamReturnCode {
    /// Successful function return
    SUCCESS = 0,

    /// dlopen() failure when dynamically loading a service module
    OPEN_ERR = 1,

    /// Symbol not found
    SYMBOL_ERR = 2,

    /// Error in service module
    SERVICE_ERR = 3,

    /// System error
    SYSTEM_ERR = 4,

    /// Memory buffer error
    BUF_ERR = 5,

    /// Permission denied
    PERM_DENIED = 6,

    /// Authentication failure
    AUTH_ERR = 7,

    /// Can not access authentication data due to insufficient credentials
    CRED_INSUFFICIENT = 8,

    /// Underlying authentication service can not retrieve authentication information
    AUTHINFO_UNAVAIL = 9,

    /// User not known to the underlying authentication module
    USER_UNKNOWN = 10,

    /// An authentication service has maintained a retry count which has been reached.
    /// No further retries should be attempted
    MAXTRIES = 11,

    /// New authentication token required.
    /// This is normally returned if the machine security policies require
    /// that the password should be changed beccause the password is NULL or it has aged
    NEW_AUTHTOK_REQD = 12,

    /// User account has expired
    ACCT_EXPIRED = 13,

    /// Can not make/remove an entry for the specified session
    SESSION_ERR = 14,

    /// Underlying authentication service can not retrieve user credentials unavailable
    CRED_UNAVAIL = 15,

    /// User credentials expired
    CRED_EXPIRED = 16,

    /// Failure setting user credentials
    CRED_ERR = 17,

    /// No module specific data is present
    NO_MODULE_DATA = 18,

    /// Conversation error
    CONV_ERR = 19,

    /// Authentication token manipulation error
    AUTHTOK_ERR = 20,

    /// Authentication information cannot be recovered
    AUTHTOK_RECOVERY_ERR = 21,

    /// Authentication token lock busy
    AUTHTOK_LOCK_BUSY = 22,

    /// Authentication token aging disabled
    AUTHTOK_DISABLE_AGING = 23,

    /// Preliminary check by password service
    TRY_AGAIN = 24,

    /// Ignore underlying account module regardless of whether
    /// the control flag is required, optional, or sufficient
    IGNORE = 25,

    /// Critical error (?module fail now request)
    AUTHTOK_EXPIRED = 27,

    /// user's authentication token has expired
    ABORT = 26,

    /// module is not known
    MODULE_UNKNOWN = 28,

    /// Bad item passed to pam_*_item()
    BAD_ITEM = 29,

    /// conversation function is event driven and data is not available yet
    CONV_AGAIN = 30,

    /// please call this function again to complete authentication stack.
    /// Before calling again, verify that conversation is completed
    INCOMPLETE = 31,
}

impl Display for PamReturnCode {
    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
        f.write_str(&format!("{:?} ({})", self, *self as i32))
    }
}

impl From<i32> for PamReturnCode {
    fn from(status: i32) -> PamReturnCode {
        match status {
            0 => PamReturnCode::SUCCESS,
            1 => PamReturnCode::OPEN_ERR,
            2 => PamReturnCode::SYMBOL_ERR,
            3 => PamReturnCode::SERVICE_ERR,
            4 => PamReturnCode::SYSTEM_ERR,
            5 => PamReturnCode::BUF_ERR,
            6 => PamReturnCode::PERM_DENIED,
            7 => PamReturnCode::AUTH_ERR,
            8 => PamReturnCode::CRED_INSUFFICIENT,
            9 => PamReturnCode::AUTHINFO_UNAVAIL,
            10 => PamReturnCode::USER_UNKNOWN,
            11 => PamReturnCode::MAXTRIES,
            12 => PamReturnCode::NEW_AUTHTOK_REQD,
            13 => PamReturnCode::ACCT_EXPIRED,
            14 => PamReturnCode::SESSION_ERR,
            15 => PamReturnCode::CRED_UNAVAIL,
            16 => PamReturnCode::CRED_EXPIRED,
            17 => PamReturnCode::CRED_ERR,
            18 => PamReturnCode::NO_MODULE_DATA,
            19 => PamReturnCode::CONV_ERR,
            20 => PamReturnCode::AUTHTOK_ERR,
            21 => PamReturnCode::AUTHTOK_RECOVERY_ERR,
            22 => PamReturnCode::AUTHTOK_LOCK_BUSY,
            23 => PamReturnCode::AUTHTOK_DISABLE_AGING,
            24 => PamReturnCode::TRY_AGAIN,
            25 => PamReturnCode::IGNORE,
            26 => PamReturnCode::ABORT,
            27 => PamReturnCode::AUTHTOK_EXPIRED,
            28 => PamReturnCode::MODULE_UNKNOWN,
            29 => PamReturnCode::BAD_ITEM,
            30 => PamReturnCode::CONV_AGAIN,
            31 => PamReturnCode::INCOMPLETE,
            _ => PamReturnCode::SYSTEM_ERR,
        }
    }
}

/// The Linux-PAM flags
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum PamFlag {
    /// Authentication service should not generate any messages
    SILENT = 0x8000,

    /// The authentication service should return AUTH_ERROR
    /// if the user has a null authentication token
    /// (used by pam_authenticate{,_secondary}())
    DISALLOW_NULL_AUTHTOK = 0x0001,

    /// Set user credentials for an authentication service
    /// (used for pam_setcred())
    ESTABLISH_CRED = 0x0002,

    /// Delete user credentials associated with an authentication service
    /// (used for pam_setcred())
    DELETE_CRED = 0x0004,

    /// Reinitialize user credentials
    /// (used for pam_setcred())
    REINITIALIZE_CRED = 0x0008,

    /// Extend lifetime of user credentials
    /// (used for pam_setcred())
    REFRESH_CRED = 0x0010,

    /// The password service should only update those passwords that have aged.
    /// If this flag is not passed, the password service should update all passwords.
    /// (used by pam_chauthtok)
    CHANGE_EXPIRED_AUTHTOK = 0x0020,

    NONE = 0x0000,
}

impl Display for PamFlag {
    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
        f.write_str(&format!("{:?} ({})", self, *self as i32))
    }
}

/// The Linux-PAM item types
///
/// These defines are used by `pam_set_item()` `and pam_get_item()`.
/// Please check the spec which are allowed for use by applications
/// and which are only allowed for use by modules.
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum PamItemType {
    /// The service name
    SERVICE = 1,

    /// The user name
    USER = 2,

    /// The tty name
    TTY = 3,

    /// The remote host name
    RHOST = 4,

    /// The pam_conv structure
    CONV = 5,

    /// The authentication token (password)
    AUTHTOK = 6,

    /// The old authentication token
    OLDAUTHTOK = 7,

    /// The remote user name
    RUSER = 8,

    /// the prompt for getting a username Linux-PAM extensions
    USER_PROMPT = 9,

    /// app supplied function to override failure delays
    FAIL_DELAY = 10,

    /// X display name
    XDISPLAY = 11,

    /// X server authentication data
    XAUTHDATA = 12,

    /// The type for pam_get_authtok
    AUTHTOK_TYPE = 13,
}

impl Display for PamItemType {
    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
        f.write_str(&format!("{:?} ({})", self, *self as i32))
    }
}

/// The Linux-PAM message styles
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum PamMessageStyle {
    PROMPT_ECHO_OFF = 1,
    PROMPT_ECHO_ON = 2,
    ERROR_MSG = 3,
    TEXT_INFO = 4,
}

impl Display for PamMessageStyle {
    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
        f.write_str(&format!("{:?} ({})", self, *self as i32))
    }
}

impl From<i32> for PamMessageStyle {
    fn from(style: i32) -> PamMessageStyle {
        match style {
            1 => PamMessageStyle::PROMPT_ECHO_OFF,
            2 => PamMessageStyle::PROMPT_ECHO_ON,
            3 => PamMessageStyle::ERROR_MSG,
            4 => PamMessageStyle::TEXT_INFO,
            _ => PamMessageStyle::ERROR_MSG,
        }
    }
}
